src/shop_type.cpp

```diff
@@ -281,14 +303,18 @@ void ShopInstance::Load(Loader &ldr)
 	SmallRideInstance inst_number = static_cast<SmallRideInstance>(this->GetIndex());
 	uint8 entrances = this->GetEntranceDirections(this->vox_pos);
 
-	Voxel *v = _world.GetCreateVoxel(this->vox_pos, true);
-	if (v != nullptr && v->GetInstance() == SRI_FREE) {
-		v->SetInstance(inst_number);
-		v->SetInstanceData(entrances);
-
-		AddRemovePathEdges(this->vox_pos, PATH_EMPTY, entrances, PAS_QUEUE_PATH);
-	} else {
-		ldr.SetFailMessage("Invalid world coordinates for shop.");
+	const int16 height = this->GetShopType()->height;
+	for (int16 i = 0; i < height; ++i) {
+		Voxel *v = _world.GetCreateVoxel(this->vox_pos + XYZPoint16(0, 0, i), true);
+		if (v != nullptr && v->GetInstance() == SRI_FREE) {
+			v->SetInstance(inst_number);
+			v->SetInstanceData(i > 0 ? INVISIBLE_SHOP_TILE : entrances);
+
+			AddRemovePathEdges(this->vox_pos, PATH_EMPTY, entrances, PAS_QUEUE_PATH);
```

Comment:
You're running this more than once now (if `height > 1`), but it doesn't seem to depend on loop data `i` or `v`.

Reply:
fixed